src/binarylane/config/sources.py

```diff
@@ -0,0 +1,130 @@
+from __future__ import annotations
+
+import argparse
+import configparser
+import os
+import sys
+from abc import ABC
+from pathlib import Path
+from typing import ClassVar, Dict, Optional
+
+
+class _SourceBase(ABC):
+    _config: Dict[str, str]
+
+    def get(self, name: str) -> Optional[str]:
+        return self._config.get(name, None)
+
+
+class CommandlineSource(_SourceBase):
+    def __init__(self, config: argparse.Namespace) -> None:
+        self._config = {key.replace("_", "-"): str(value) for key, value in vars(config).items() if value is not None}
+
+
+class DefaultSource(_SourceBase):
+    def __init__(self) -> None:
+        self._config = {
+            "api-url": "https://api.binarylane.com.au",
+            "context": "bl",
+        }
+
+
+class EnvironmentSource(_SourceBase):
+    prefix: ClassVar[str] = "BL_"
+
+    def __init__(self) -> None:
+        self._config = {self._get_name(key): value for key, value in os.environ.items() if key.startswith(self.prefix)}
+
+    def _get_name(self, key: str) -> str:
+        return key[len(self.prefix) :].lower().replace("_", "-")
+
+
+class FileSource:
+    DEFAULT_PATH = Path("__DEFAULT__")
+    _DIRNAME = "binarylane"
+    _FILENAME = "config.ini"
+    _path: Optional[Path]
+    section_name: str
+
+    _parser: configparser.ConfigParser
+
+    def __init__(self, config_file: Optional[Path] = None) -> None:
+        if config_file is self.DEFAULT_PATH:
+            config_file = self._get_config_dir() / self._FILENAME
+        self._path = config_file
+
+        self._parser = configparser.ConfigParser()
+        self._read()
+        self.section_name = configparser.DEFAULTSECT
+
+    @staticmethod
+    def _get_config_home() -> Path:
+        """Return platform-specific path that programs should store configuration in"""
+        # On windows, configuration is stored in APPDATA
+        if sys.platform == "win32":
+            appdata = os.getenv("APPDATA")
+            if not appdata:
+                raise EnvironmentError("%APPDATA% is not set?")
+            return Path(appdata)
+
+        # On other systems, use XDG_CONFIG_HOME if set
+        xdg_config_home = os.getenv("XDG_CONFIG_HOME")
+        if xdg_config_home:
+            return Path(xdg_config_home)
+
+        # Otherwise, use $HOME/.config
+        home = os.getenv("HOME")
```

Comment:
We could also use  [Path.home()](https://docs.python.org/3/library/pathlib.html#pathlib.Path.home) here which works by expanding `~`, not sure if that is more "cross platform" than this though

Reply:
Was not aware of `Path.home()`, I will stdlib source to see how it works - but at face value, agree that seems the better solution.